lib/utils/getFormatter.mjs

```diff
@@ -0,0 +1,57 @@
+import { EXIT_CODE_INVALID_CONFIG } from '../constants.mjs';
+import formatters from '../formatters/index.mjs';
+import getConfigForFile from '../getConfigForFile.mjs';
+import getFormatterOptionsText from './getFormatterOptionsText.mjs';
+
+/** @import {Formatter, FormatterType, InternalApi} from 'stylelint' */
+
+/**
+ * @param {InternalApi} stylelint
+ * @param {Object} [params]
+ * @param {FormatterType | Formatter | undefined} [params.formatterType]
+ * @returns {Promise<Formatter>}
+ */
+export default async function getFormatter(stylelint) {
+	const cwd = stylelint._options.cwd;
+	const configPath = stylelint._options.configFile || cwd;
+	let formatter = stylelint._options.formatter;
+
+	if (!formatter) {
+		let configForFile;
+
+		try {
+			configForFile = await getConfigForFile(stylelint, configPath);
+		} catch (err) {
+			if (err instanceof Error && 'code' in err && err.code === EXIT_CODE_INVALID_CONFIG) {
+				configForFile = undefined;
+			} else {
+				throw err;
+			}
+		}
+
+		const resolvedConfig = stylelint._options.config ? stylelint._options : configForFile;
+
+		formatter = resolvedConfig?.config.formatter;
+	}
+
+	if (typeof formatter === 'string') {
+		const formatterFunction = formatters[formatter];
+
+		if (formatterFunction === undefined) {
+			const formattersText = getFormatterOptionsText(', ', '"');
+
+			throw new Error(`You must use a valid formatter option: ${formattersText} or a function`);
+		}
+
+		return formatterFunction;
+	}
+
+	// Assume a function or a promise of a function.
+	if (typeof formatter === 'function' || formatter) {
+		return Promise.resolve(formatter);
+	}
+
+	formatter ??= stylelint._options._defaultFormatter ?? 'json';
```

Comment:
question

Is `_defaultFormatter` only being used here?
If so what about removing it completely and just using `DEFAULT_FORMATTER` directly here?

Reply:
Sadyly we cannot do that because `standalone` uses `json` by default, while `cli` uses `string` by default. The default formatter depends on the context.

Reply:
@ybiquitous [this line](https://github.com/stylelint/stylelint/pull/7826/files#diff-d1470268bb8c3a42692b2c5249d7d31832ceeed1b1f16fbdbc814f8013c36b31R379) is the only line where `_defaultFormatter` is set, correct?
Are you saying that it's necessary so that it can be set in our tests?

Reply:
Yes, `_defaultFormatter` is set only in `cli` for now. This is not for the test but for CLI. Users shouldn't use it, but the Stylelint implementation needs it.

Reply:
> **for now**

As I understand it, `_defaultFormatter` should be set elsewhere; does it mean that the context switch will be done in this PR?

Reply:
No more changes are necessary in this PR, I think.

Reply:
> but the Stylelint implementation needs it

What I am trying to get at is _where/when_?
i.e. if `standalone` uses `'json'` by default, where or when is `_defaultFormatter` supposed to be set to `'json'`?
In a separate PR?
If it's not meant to be used for our tests nor be used by users, then what will set it to `'json'`?